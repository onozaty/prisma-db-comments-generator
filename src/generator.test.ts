import * as child_process from "node:child_process";
import fs from "node:fs";
import path from "node:path";

const fixturesDir = path.join(__dirname, "__fixtures__");

beforeAll(() => {
  // テスト開始時にビルドしておく
  child_process.execSync(`pnpm run build`);
});

beforeEach(() => {
  child_process.execSync(
    `find ${fixturesDir} -name "migrations" -type d -prune -exec rm -rf {} \\;`,
  );
});

test("basic", async () => {
  // Arrange
  const name = "basic";

  // Act
  executeGenerate(name);

  // Assert
  const migrationSqlContent = readMigrationSql(name);
  expect(migrationSqlContent).toMatchSnapshot("migration.sql");

  const commentsLatestJsonContent = readCommentsLatestJson(name);
  expect(commentsLatestJsonContent).toMatchSnapshot("comments-latest.json");
});

test("target-column", async () => {
  // Arrange
  const name = "target-column";

  // Act
  executeGenerate(name);

  // Assert
  const migrationSqlContent = readMigrationSql(name);
  expect(migrationSqlContent).toMatchSnapshot("migration.sql");

  const commentsLatestJsonContent = readCommentsLatestJson(name);
  expect(commentsLatestJsonContent).toMatchSnapshot("comments-latest.json");
});

test("target-table", async () => {
  // Arrange
  const name = "target-table";

  // Act
  executeGenerate(name);

  // Assert
  const migrationSqlContent = readMigrationSql(name);
  expect(migrationSqlContent).toMatchSnapshot("migration.sql");

  const commentsLatestJsonContent = readCommentsLatestJson(name);
  expect(commentsLatestJsonContent).toMatchSnapshot("comments-latest.json");
});

test("diff", async () => {
  // Arrange
  // diff1のマイグレーションをベースに、diff2のマイグレーションを生成して差分となること
  executeGenerate("diff1");
  fs.cpSync(getMigrationsDir("diff1"), getMigrationsDir("diff2"), {
    recursive: true,
  });

  const name = "diff2";

  // Act
  executeGenerate(name);

  // Assert
  const migrationSqlContent = readMigrationSql(name);
  expect(migrationSqlContent).toMatchSnapshot("migration.sql");

  const commentsLatestJsonContent = readCommentsLatestJson(name);
  expect(commentsLatestJsonContent).toMatchSnapshot("comments-latest.json");
});

test("multi-schema", async () => {
  // Arrange
  const name = "multi-schema";

  // Act
  executeGenerate(name);

  // Assert
  const migrationSqlContent = readMigrationSql(name);
  expect(migrationSqlContent).toMatchSnapshot("migration.sql");

  const commentsLatestJsonContent = readCommentsLatestJson(name);
  expect(commentsLatestJsonContent).toMatchSnapshot("comments-latest.json");
});

test("mysql", async () => {
  // Arrange
  const name = "mysql";

  // Act
  executeGenerate(name);

  // Assert
  const migrationSqlContent = readMigrationSql(name);
  expect(migrationSqlContent).toMatchSnapshot("migration.sql");

  const commentsLatestJsonContent = readCommentsLatestJson(name);
  expect(commentsLatestJsonContent).toMatchSnapshot("comments-latest.json");
});

const executeGenerate = (name: string) => {
  const schemaPath = path.join(fixturesDir, name, "schema.prisma");
  child_process.execSync(`npx prisma generate --schema ${schemaPath}`);
};

const readCommentsLatestJson = (name: string): string => {
  const migrationsDir = getMigrationsDir(name);
  const commentLatestJsonPath = path.join(
    migrationsDir,
    "comments-latest.json",
  );

  return fs.readFileSync(commentLatestJsonPath, "utf-8");
};

const readMigrationSql = (name: string): string => {
  const migrationsDir = getMigrationsDir(name);
  const latestDirName = fs
    .readdirSync(migrationsDir, { withFileTypes: true })
    .filter((dirent) => dirent.isDirectory())
    .map((dirent) => dirent.name)
    .sort()
    .reverse()[0];

  if (!latestDirName) {
    throw new Error("No migration directories found");
  }

  const migrationSqlPath = path.join(
    migrationsDir,
    latestDirName,
    "migration.sql",
  );

  const contents = fs.readFileSync(migrationSqlPath, "utf-8");

  // Remove version-dependent parts (e.g., "-- Prisma Database Comments Generator vX.Y.Z") from the migration SQL content.
  return contents.replace(
    /-- Prisma Database Comments Generator v[0-9]+\.[0-9]+\.[0-9]+\n/,
    "",
  );
};

const getMigrationsDir = (name: string) => {
  return path.join(fixturesDir, name, "migrations");
};
