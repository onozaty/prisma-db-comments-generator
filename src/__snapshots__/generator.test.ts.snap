// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`basic > comments-latest.json 1`] = `
"{
  "customers": {
    "table": {
      "tableName": "customers",
      "comment": "Customer"
    },
    "columns": [
      {
        "tableName": "customers",
        "columnName": "customer_id",
        "comment": "Customer ID"
      },
      {
        "tableName": "customers",
        "columnName": "name",
        "comment": "Customer Name"
      },
      {
        "tableName": "customers",
        "columnName": "email",
        "comment": "e-mail"
      },
      {
        "tableName": "customers",
        "columnName": "created_at",
        "comment": ""
      }
    ]
  },
  "products": {
    "table": {
      "tableName": "products",
      "comment": "Product"
    },
    "columns": [
      {
        "tableName": "products",
        "columnName": "product_id",
        "comment": "Product ID"
      },
      {
        "tableName": "products",
        "columnName": "name",
        "comment": "Product Name"
      },
      {
        "tableName": "products",
        "columnName": "type",
        "comment": "Product Type\\nenum: enum_product_type(BOOK, TOY, FASHION)"
      },
      {
        "tableName": "products",
        "columnName": "description",
        "comment": "Product Description"
      },
      {
        "tableName": "products",
        "columnName": "price",
        "comment": "Price"
      },
      {
        "tableName": "products",
        "columnName": "created_at",
        "comment": ""
      }
    ]
  },
  "sales": {
    "table": {
      "tableName": "sales",
      "comment": ""
    },
    "columns": [
      {
        "tableName": "sales",
        "columnName": "sale_id",
        "comment": "Sale ID"
      },
      {
        "tableName": "sales",
        "columnName": "customer_id",
        "comment": "Customer ID"
      },
      {
        "tableName": "sales",
        "columnName": "product_id",
        "comment": "Product ID"
      },
      {
        "tableName": "sales",
        "columnName": "quantity",
        "comment": "Quantity"
      },
      {
        "tableName": "sales",
        "columnName": "total_price",
        "comment": "Total Price"
      },
      {
        "tableName": "sales",
        "columnName": "created_at",
        "comment": ""
      }
    ]
  }
}"
`;

exports[`basic > migration.sql 1`] = `
"
-- customers comments
COMMENT ON TABLE "customers" IS 'Customer';
COMMENT ON COLUMN "customers"."customer_id" IS 'Customer ID';
COMMENT ON COLUMN "customers"."name" IS 'Customer Name';
COMMENT ON COLUMN "customers"."email" IS 'e-mail';

-- products comments
COMMENT ON TABLE "products" IS 'Product';
COMMENT ON COLUMN "products"."product_id" IS 'Product ID';
COMMENT ON COLUMN "products"."name" IS 'Product Name';
COMMENT ON COLUMN "products"."type" IS E'Product Type\\nenum: enum_product_type(BOOK, TOY, FASHION)';
COMMENT ON COLUMN "products"."description" IS 'Product Description';
COMMENT ON COLUMN "products"."price" IS 'Price';

-- sales comments
COMMENT ON COLUMN "sales"."sale_id" IS 'Sale ID';
COMMENT ON COLUMN "sales"."customer_id" IS 'Customer ID';
COMMENT ON COLUMN "sales"."product_id" IS 'Product ID';
COMMENT ON COLUMN "sales"."quantity" IS 'Quantity';
COMMENT ON COLUMN "sales"."total_price" IS 'Total Price';
"
`;

exports[`diff > comments-latest.json 1`] = `
"{
  "products": {
    "table": {
      "tableName": "products",
      "comment": ""
    },
    "columns": [
      {
        "tableName": "products",
        "columnName": "product_id",
        "comment": "Product ID"
      },
      {
        "tableName": "products",
        "columnName": "name",
        "comment": "Product Name"
      },
      {
        "tableName": "products",
        "columnName": "type",
        "comment": "Product Type\\nenum: enum_product_type(BOOK, TOY, FASHION)"
      },
      {
        "tableName": "products",
        "columnName": "description",
        "comment": "Product Description"
      },
      {
        "tableName": "products",
        "columnName": "price",
        "comment": "Price"
      },
      {
        "tableName": "products",
        "columnName": "created_at",
        "comment": ""
      },
      {
        "tableName": "products",
        "columnName": "shopShopId",
        "comment": ""
      }
    ]
  },
  "sales": {
    "table": {
      "tableName": "sales",
      "comment": "Sale"
    },
    "columns": [
      {
        "tableName": "sales",
        "columnName": "sale_id",
        "comment": "Sale ID"
      },
      {
        "tableName": "sales",
        "columnName": "product_id",
        "comment": ""
      },
      {
        "tableName": "sales",
        "columnName": "quantity",
        "comment": "Quantity"
      },
      {
        "tableName": "sales",
        "columnName": "total_price",
        "comment": "Total Price X"
      },
      {
        "tableName": "sales",
        "columnName": "created_at",
        "comment": ""
      }
    ]
  },
  "shops": {
    "table": {
      "tableName": "shops",
      "comment": "Shop"
    },
    "columns": [
      {
        "tableName": "shops",
        "columnName": "shop_id",
        "comment": "Shop ID"
      },
      {
        "tableName": "shops",
        "columnName": "name",
        "comment": "Shop Name"
      },
      {
        "tableName": "shops",
        "columnName": "created_at",
        "comment": "Created At"
      }
    ]
  }
}"
`;

exports[`diff > migration.sql 1`] = `
"
-- products comments
COMMENT ON TABLE "products" IS NULL;

-- sales comments
COMMENT ON TABLE "sales" IS 'Sale';
COMMENT ON COLUMN "sales"."product_id" IS NULL;
COMMENT ON COLUMN "sales"."total_price" IS 'Total Price X';

-- shops comments
COMMENT ON TABLE "shops" IS 'Shop';
COMMENT ON COLUMN "shops"."shop_id" IS 'Shop ID';
COMMENT ON COLUMN "shops"."name" IS 'Shop Name';
COMMENT ON COLUMN "shops"."created_at" IS 'Created At';
"
`;

exports[`multi-schema > comments-latest.json 1`] = `
"{
  "foo.registered_user": {
    "table": {
      "schema": "foo",
      "tableName": "registered_user",
      "comment": "A registered user"
    },
    "columns": [
      {
        "schema": "foo",
        "tableName": "registered_user",
        "columnName": "id",
        "comment": "Primary key"
      },
      {
        "schema": "foo",
        "tableName": "registered_user",
        "columnName": "username",
        "comment": "Username used to login"
      },
      {
        "schema": "foo",
        "tableName": "registered_user",
        "columnName": "firstname",
        "comment": "User's firstname"
      },
      {
        "schema": "foo",
        "tableName": "registered_user",
        "columnName": "lastname",
        "comment": "User's lastname"
      },
      {
        "schema": "foo",
        "tableName": "registered_user",
        "columnName": "email",
        "comment": "Email address"
      }
    ]
  }
}"
`;

exports[`multi-schema > migration.sql 1`] = `
"
-- foo.registered_user comments
COMMENT ON TABLE "foo"."registered_user" IS 'A registered user';
COMMENT ON COLUMN "foo"."registered_user"."id" IS 'Primary key';
COMMENT ON COLUMN "foo"."registered_user"."username" IS 'Username used to login';
COMMENT ON COLUMN "foo"."registered_user"."firstname" IS 'User''s firstname';
COMMENT ON COLUMN "foo"."registered_user"."lastname" IS 'User''s lastname';
COMMENT ON COLUMN "foo"."registered_user"."email" IS 'Email address';
"
`;

exports[`target-column > comments-latest.json 1`] = `
"{
  "customers": {
    "columns": [
      {
        "tableName": "customers",
        "columnName": "customer_id",
        "comment": "Customer ID"
      },
      {
        "tableName": "customers",
        "columnName": "name",
        "comment": "Customer Name"
      },
      {
        "tableName": "customers",
        "columnName": "email",
        "comment": "e-mail"
      },
      {
        "tableName": "customers",
        "columnName": "created_at",
        "comment": ""
      }
    ]
  },
  "products": {
    "columns": [
      {
        "tableName": "products",
        "columnName": "product_id",
        "comment": "Product ID"
      },
      {
        "tableName": "products",
        "columnName": "name",
        "comment": "Product Name"
      },
      {
        "tableName": "products",
        "columnName": "type",
        "comment": "Product Type"
      },
      {
        "tableName": "products",
        "columnName": "description",
        "comment": "Product Description"
      },
      {
        "tableName": "products",
        "columnName": "price",
        "comment": "Price"
      },
      {
        "tableName": "products",
        "columnName": "created_at",
        "comment": ""
      }
    ]
  },
  "sales": {
    "columns": [
      {
        "tableName": "sales",
        "columnName": "sale_id",
        "comment": "Sale ID"
      },
      {
        "tableName": "sales",
        "columnName": "customer_id",
        "comment": "Customer ID"
      },
      {
        "tableName": "sales",
        "columnName": "product_id",
        "comment": "Product ID"
      },
      {
        "tableName": "sales",
        "columnName": "quantity",
        "comment": "Quantity"
      },
      {
        "tableName": "sales",
        "columnName": "total_price",
        "comment": "Total Price"
      },
      {
        "tableName": "sales",
        "columnName": "created_at",
        "comment": ""
      }
    ]
  }
}"
`;

exports[`target-column > migration.sql 1`] = `
"
-- customers comments
COMMENT ON COLUMN "customers"."customer_id" IS 'Customer ID';
COMMENT ON COLUMN "customers"."name" IS 'Customer Name';
COMMENT ON COLUMN "customers"."email" IS 'e-mail';

-- products comments
COMMENT ON COLUMN "products"."product_id" IS 'Product ID';
COMMENT ON COLUMN "products"."name" IS 'Product Name';
COMMENT ON COLUMN "products"."type" IS 'Product Type';
COMMENT ON COLUMN "products"."description" IS 'Product Description';
COMMENT ON COLUMN "products"."price" IS 'Price';

-- sales comments
COMMENT ON COLUMN "sales"."sale_id" IS 'Sale ID';
COMMENT ON COLUMN "sales"."customer_id" IS 'Customer ID';
COMMENT ON COLUMN "sales"."product_id" IS 'Product ID';
COMMENT ON COLUMN "sales"."quantity" IS 'Quantity';
COMMENT ON COLUMN "sales"."total_price" IS 'Total Price';
"
`;

exports[`target-table > comments-latest.json 1`] = `
"{
  "customers": {
    "table": {
      "tableName": "customers",
      "comment": "Customer"
    }
  },
  "products": {
    "table": {
      "tableName": "products",
      "comment": "Product"
    }
  },
  "sales": {
    "table": {
      "tableName": "sales",
      "comment": ""
    }
  }
}"
`;

exports[`target-table > migration.sql 1`] = `
"
-- customers comments
COMMENT ON TABLE "customers" IS 'Customer';

-- products comments
COMMENT ON TABLE "products" IS 'Product';
"
`;
